### ISSUE TITLE
Logs are not threaded together on Go 1.11 standard environment

### ISSUE THREAD

**Issue by mehdi-sol** [andyfusniak/stackdriver-gae-logrus-plugin#1]
Hi,

I want to imitate [appengine/log](https://godoc.org/google.golang.org/appengine/log) using this library. While I am currently on Go 1.11 I'm doing this to be prepared to make the move to Go 1.12+. Here's my `log` package.

```
package log

import (
	"context"
	stackdriver "github.com/andyfusniak/stackdriver-gae-logrus-plugin"
	log "github.com/sirupsen/logrus"
	"os"
)

func Init() {
	projectID := os.Getenv("GOOGLE_CLOUD_PROJECT")
	if projectID == "" {
		projectID = "some_project_id"
		log.Printf("Defaulting to project ID %s", projectID)
	}
	formatter := stackdriver.GAEStandardFormatter(
		stackdriver.WithProjectID(projectID),
	)
	log.SetFormatter(formatter)
	log.SetOutput(os.Stdout)
}

func Debugf(ctx context.Context, format string, args ...interface{}) {
	log.WithContext(ctx).Debugf(format, args)
}

func Infof(ctx context.Context, format string, args ...interface{}) {
	log.WithContext(ctx).Infof(format, args)
}

func Errorf(ctx context.Context, format string, args ...interface{}) {
	log.WithContext(ctx).Errorf(format, args)
}
```

While I can confirm logging in fact works and logs via Errorf, Infof and Debugf show up under GAE logs, they are not nested in their respective request logs and show up as a different log. This is different from what appengine/log does, which makes any applications a child of their respective request log. 

Any help would be appreciated.

**Comment by andyfusniak** [comment]
To thread the log messages you will need to extract the trace identifier from the `X-Cloud-Trace-Context` header:

From this page: https://cloud.google.com/appengine/docs/standard/go/writing-application-logs#writing_app_logs

> If you want to correlate entries in the app log with the request log, your structured app log entries need to contain the request's trace identifier. You can extract the trace identifier from the X-Cloud-Trace-Context request header. In your structured log entry, write the ID to a field named logging.googleapis.com/trace. For more information about the X-Cloud-Trace-Context header, see Forcing a request to be traced.

Inside this repository I wrote an example Go HTTP middleware component `XCloudTraceContext`
https://github.com/andyfusniak/stackdriver-gae-logrus-plugin/blob/master/middleware/x_cloud_trace_context.go

You'll need to get this middleware to run before handling HTTP requests that intend to write logs. This line https://github.com/andyfusniak/stackdriver-gae-logrus-plugin/blob/master/middleware/x_cloud_trace_context.go#L31
 will inject the trace id into the context under the `xctc` key.

In the logging code (https://github.com/andyfusniak/stackdriver-gae-logrus-plugin/blob/master/formatter.go#L132) is a check to see if this has been set. If set, it will setup the appropriate trace string for the structured logging. This in turn should cause Google's log viewer to display the logs in a threaded fashion.

In summary, introduce the `XCloudTraceContext` middleware and this should solve your issue. Let me know how it goes.

**Comment by mehdi-sol** [comment]
Thank you Andy.

This does not seem to fix the issue. I'm using Gorilla mux and here's the middleware:

```
func loggingMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		ctx := context.WithValue(r.Context(), "xctc", r.Header.Get("X-Cloud-Trace-Context"))
		next.ServeHTTP(w, r.WithContext(ctx))
	})
}
```

As mentioned in Gorilla mux [documentation](https://github.com/gorilla/mux#middleware), I'm using this middleware like this:
```
r := mux.NewRouter()
r.HandleFunc("/", api.IndexPage).Methods("GET")
...
...
r.Use(loggingMiddleware)
```
Also `r.Use(middleware.XCloudTraceContext)` didn't seem to work. 

Any ideas what might be the issue?

**Comment by andyfusniak** [comment]
I can only suggest logging the `GOOGLE_CLOUD_PROJECT` and `r.Header.Get("X-Cloud-Trace-Context")` values and double checking them both. Perform a GET request and check your log stream to ensure they are indeed set, and if so to an appropriate value?

**Comment by mehdi-sol** [comment]
I can confirm both `GOOGLE_CLOUD_PROJECT` environment variable is set and request headers do infact include "X-Cloud-Trace-Context".

```
func loggingMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		xctc := r.Header.Get("X-Cloud-Trace-Context")
		if xctc == "" {
			log.Print("XCTC header not set")
		} else {
			log.Printf("XCTC is set and is %v", xctc)
		}
		proj_name := os.Getenv("GOOGLE_CLOUD_PROJECT")
		if proj_name == "" {
			log.Print("Project name is not set")
		} else {
			log.Printf("Porject name is set and is %v", proj_name)
		}
		ctx := context.WithValue(r.Context(), "xctc", r.Header.Get("X-Cloud-Trace-Context"))
		next.ServeHTTP(w, r.WithContext(ctx))
	})
}
```

Here are the relating logs from stackdriver:
`textPayload: "2020/07/16 18:28:01 XCTC is set and is b58cb6607b1030f4e74196912cd100c6/3152785209952454850;o=1"  `
`textPayload: "2020/07/16 18:28:01 Porject name is set and is hotline_bling_beta"  `

**Comment by mehdi-sol** [comment]
Never mind. using `r.Use(middleware.XCloudTraceContext)` worked. Thank you.
